src/Algebra/Graph.hs

```diff
@@ -232,49 +234,51 @@ instance Num a => Num (Graph a) where
     abs         = id
     negate      = id
 
+-- | `==` is a good consumer of both arguments.
 instance Ord a => Eq (Graph a) where
     (==) = eqR
 
+-- | 'compare' is a good consumer of both arguments.
 instance Ord a => Ord (Graph a) where
     compare = ordR
 
 -- TODO: Find a more efficient equality check.
 -- Check if two graphs are equal by converting them to their adjacency maps.
 eqR :: Ord a => Graph a -> Graph a -> Bool
 eqR x y = toAdjacencyMap x == toAdjacencyMap y
-{-# NOINLINE [1] eqR #-}
+{-# INLINE [2] eqR #-}
 {-# RULES "eqR/Int" eqR = eqIntR #-}
 
 -- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
 eqIntR :: Graph Int -> Graph Int -> Bool
-eqIntR x y = toAdjacencyIntMap x == toAdjacencyIntMap y
+eqIntR = \x y -> toAdjacencyIntMap x == toAdjacencyIntMap y
+{-# INLINE eqIntR #-}
 
 -- TODO: Find a more efficient comparison.
 -- Compare two graphs by converting them to their adjacency maps.
 ordR :: Ord a => Graph a -> Graph a -> Ordering
 ordR x y = compare (toAdjacencyMap x) (toAdjacencyMap y)
-{-# NOINLINE [1] ordR #-}
+{-# INLINE [2] ordR #-}
 {-# RULES "ordR/Int" ordR = ordIntR #-}
 
 -- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
 ordIntR :: Graph Int -> Graph Int -> Ordering
-ordIntR x y = compare (toAdjacencyIntMap x) (toAdjacencyIntMap y)
+ordIntR = \x y -> compare (toAdjacencyIntMap x) (toAdjacencyIntMap y)
+{-# INLINE ordIntR #-}
 
+-- | `<*>` is a good consumer of its first agument and producer.
```

Comment:
Is it possible to make it a good consumer of both arguments? If not, is there a simple argument why this can't be done?

Reply:
(If there is such an argument, I think it'd be useful to add it above as a comment.)

Reply:
There is no arguments, I don't get it. Inspecting core, the second argument is hidden in a local variable until last phase, preventing the rule to fire... I spent some time without success, so I will add a comment for now.